### DISCUSSION TITLE
Marker value with text

### DISCUSSION THREAD

**Discussion by EnrikasVai** [patrykandpatrick/vico#628] [Questions] [answered]
Hello, I'm using the Vico chart library for the first time and can't find where to change the marker output. I copied the sample line chart added my data, customized the bottom axis and so on. The problem I have encountered is that when i press on the chart it shows me a marker with the value from the chart but I need to add custom text. For example the marker should show 78 BPM Average, but by default shows only the value 78.00  
![Screenshot_20240324-183338_bakis](https://github.com/patrykandpatrick/vico/assets/135996448/f3a974fc-8fed-4ad6-90dc-7e88e9070b2b)

**Comment by Gowsky** [comment]
Hello, see [here](https://patrykandpatrick.com/vico/api/2.0.0-alpha.12/core/com.patrykandpatrick.vico.core.component.marker/-marker-component/#1833179916%2FProperties%2F186187275).

**Reply by EnrikasVai** [reply]
Could you provide an example for me? I need to override the getLabel?

**Reply by Gowsky** [reply]
For this use case, it's enough to use [`DefaultMarkerLabelFormatter`](https://patrykandpatrick.com/vico/api/2.0.0-alpha.12/core/com.patrykandpatrick.vico.core.marker/-default-marker-label-formatter/). See `decimalFormat` argument in the constructor. 

I'd also highly recommend reading about [interfaces](https://kotlinlang.org/docs/interfaces.html).

**Reply by EnrikasVai** [reply]
Ended up using this 

```
class CustomMarkerLabelFormatter(
    private val dataset: List<Float>,
    private val markerText: String
) : MarkerLabelFormatter {
    override fun getLabel(
        markedEntries: List<Marker.EntryModel>,
        chartValues: ChartValues
    ): CharSequence {
        val labelBuilder = StringBuilder()
        for (entry in markedEntries) {
            val index = entry.index
            val displayValue = String.format("%.1f", dataset[index])

            labelBuilder.append("$markerText ").append(displayValue).append("\n")
        }
        if (labelBuilder.isNotEmpty()) {
            labelBuilder.setLength(labelBuilder.length - 1)
        }
        return labelBuilder.toString()
    }
}

@Composable
internal fun rememberMarker(
    labelPosition: MarkerComponent.LabelPosition = MarkerComponent.LabelPosition.Top,
    data: List<Float>,
    markerText: String
): Marker {

    val customLabelFormatter = CustomMarkerLabelFormatter(data, markerText)

   //
    return remember(label, labelPosition, indicator, guideline, customLabelFormatter) {
        @SuppressLint("RestrictedApi")
        object : MarkerComponent(label, labelPosition, indicator, guideline) {
            init {
                labelFormatter = customLabelFormatter
                indicatorSizeDp = 36f
                onApplyEntryColor = { entryColor ->
                    indicatorRearComponent.color = entryColor.copyColor(alpha = .15f)
                    with(indicatorCenterComponent) {
                        color = entryColor
                        setShadow(radius = 12f, color = entryColor)
                    }
                }
            }
     //
}
```

**Reply by patrickmichalik** [reply, marked as answer]
Hello, @EnrikasVai! I’m glad you’ve found a solution. I’d advise against storing the chart data in the `MarkerLabelFormatter`. You can retrieve a `Marker.EntryModel`’s corresponding _y_ value directly:

```kt
(markerEntryModel.entry as LineCartesianLayerModel.Entry).y
```

**Reply by ivanleong** [reply]
@Gowsky is it possible to have text in a marker on multiple lines? When i add a new line tag in the StringBuilder, it gets elipsized, instead.

**Reply by Gowsky** [reply]
@ivanleong, yes, you have to increase the `lineCount` in the `TextComponent`.
